### PR TITLE
swarm: skip dialing WebTransport addresses when we have QUIC addresses

### PR DESCRIPTION
**Pull request by marten-seemann**
Question: Should we have something similar for WebSocket and TCP?

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
Do I understand correctly that this would drop all webtransport addresses in case there is a single quic address?

In case a node advertises the following two address, the local node should still dial both, right?

```
"/ip4/1.2.3.4/udp/443/quic
"/ip4/5.6.7.8/udp/443/quic/webtransport
```

In other words, only if the addresses are equal on everything before `quic`, only then should the QUIC address replace the webtransport address.

Does that make sense?

**Comment by marten-seemann** [comment]
> In other words, only if the addresses are equal on everything before quic, only then should the QUIC address replace the webtransport address.

Yes, that would be better, but also more code. Not sure if it's worth the complexity.

**Comment by marten-seemann** [comment]
Actually, this probably won't be too hard. I'll update the PR.

File: p2p/net/swarm/swarm_dial.go [hunk 1]
```diff
@@ -530,3 +530,41 @@ func isRelayAddr(addr ma.Multiaddr) bool {
...
+// Happy-Eyeballs-style dialing.
+func maybeRemoveWebTransportAddrs(addrs []ma.Multiaddr) []ma.Multiaddr {
+	var hasQuic, hasWebTransport bool
+	for _, addr := range addrs {
+		if isQUIC(addr) {
+			hasQuic = true
+		}
+		if isWebTransport(addr) {
+			hasWebTransport = true
+		}
```
Comment (MarcoPolo) [inline p2p/net/swarm/swarm_dial.go:556]:
```suggestion
		hasQuic = isQuic(addr)
		hasWebTransport = isWebTransport(addr)
```

**Review by MarcoPolo — APPROVED** [review]

File: p2p/net/swarm/swarm_dial.go (hunk 1 above)
Comment (marten-seemann) [reply p2p/net/swarm/swarm_dial.go]: Oh wait, this is not correct. The ifs are in a loop. Will revert.

**Comment by marten-seemann** [comment]
Merging this simple version now. We can still have better logic later (or implement the proper Happy Eyeballs fix later).

**Pull request merged**